TestPrepareCheckerWithTransferLeader is unstable

## Flaky Test

### Which jobs are failing
```
--- FAIL: TestPrepareCheckerWithTransferLeader (38.21s)
    region_syncer_test.go:294: 
        	Error Trace:	/home/runner/work/pd/pd/tests/server/region_syncer/region_syncer_test.go:294
        	Error:      	Not equal: 
        	            	expected: "pd1"
        	            	actual  : "pd2"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-pd1
        	            	+pd2
        	Test:       	TestPrepareCheckerWithTransferLeader
FAIL
```

### CI link

https://github.com/tikv/pd/actions/runs/9737924850/job/26924893445
### Reason for failure (if possible)


### Anything else

> meet again [tikv/pd/actions/runs/9762010374/job/26944878762](https://github.com/tikv/pd/actions/runs/9762010374/job/26944878762)
> 
> ```shell
> 2024-07-02T13:54:15.0492409Z --- FAIL: TestPrepareCheckerWithTransferLeader (76.09s)
> 2024-07-02T13:54:15.0492734Z     region_syncer_test.go:289: 
> 2024-07-02T13:54:15.0493797Z         	Error Trace:	/home/runner/work/pd/pd/tests/server/region_syncer/region_syncer_test.go:289
> 2024-07-02T13:54:15.0494204Z         	Error:      	Not equal: 
> 2024-07-02T13:54:15.0494642Z         	            	expected: "pd2"
> 2024-07-02T13:54:15.0495194Z         	            	actual  : ""
> 2024-07-02T13:54:15.0495495Z         	            	
> 2024-07-02T13:54:15.0496104Z         	            	Diff:
> 2024-07-02T13:54:15.0496743Z         	            	--- Expected
> 2024-07-02T13:54:15.0497390Z         	            	+++ Actual
> 2024-07-02T13:54:15.0498020Z         	            	@@ -1 +1 @@
> 2024-07-02T13:54:15.0498413Z         	            	-pd2
> 2024-07-02T13:54:15.0498676Z         	            	+
> 2024-07-02T13:54:15.0499001Z         	Test:       	TestPrepareCheckerWithTransferLeader
> 2024-07-02T13:54:15.0499178Z FAIL
> ```

I found this error is in an extreme situation.

https://github.com/tikv/pd/blob/7e18a69734f2ae391516625ec0230c05b363f5c5/tests/server/region_syncer/region_syncer_test.go#L280-L289

After resign leader to pd-2, which is happened in `2024/07/02 13:53:11.565 +00:00`
```
!!!resign triggered by API!!!
2024-07-02T13:54:14.8419029Z [2024/07/02 13:53:11.565 +00:00] [INFO] [member.go:358] ["try to resign etcd leader to next pd-server"] [from=pd1] [to=]
```
**Due to the slow disk, the leader changed 8 times in the 40s! and no leader check is in 40s which result error**
```bash
2024-07-02T13:54:14.8776867Z [2024/07/02 13:53:21.728 +00:00] [WARN] [lease.go:175] ["lease keep alive failed"] [purpose="leader election"] [start=2024/07/02 13:53:21.726 +00:00] [error="context canceled"]
2024-07-02T13:54:14.8778148Z [2024/07/02 13:53:21.734 +00:00] [INFO] [lease.go:192] ["stop lease keep alive worker"] [purpose="leader election"]
```

For this extreme situation, I think we can failed directly.


The detailed content is

```bash
2024-07-02T13:54:14.7951174Z [2024/07/02 13:53:01.139 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd1]
2024-07-02T13:54:14.8302504Z [2024/07/02 13:53:07.537 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd1]

!!!resign triggered by API!!!
2024-07-02T13:54:14.8419029Z [2024/07/02 13:53:11.565 +00:00] [INFO] [member.go:358] ["try to resign etcd leader to next pd-server"] [from=pd1] [to=]

2024-07-02T13:54:14.8523786Z [2024/07/02 13:53:12.651 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd2]
2024-07-02T13:54:14.8715386Z [2024/07/02 13:53:18.726 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd2]
2024-07-02T13:54:14.8900582Z [2024/07/02 13:53:24.777 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd2]

!!!resign pd-2 ->0!!!
2024-07-02T13:54:14.9121126Z [2024/07/02 13:53:30.827 +00:00] [INFO] [member.go:358] ["try to resign etcd leader to next pd-server"] [from=pd2] [to=]
2024-07-02T13:54:14.9167935Z [2024/07/02 13:53:31.329 +00:00] [ERROR] [server.go:1718] ["campaign PD leader meets error due to etcd error"] [campaign-leader-name=pd2] [error="[PD:server:ErrLeaderFrequentlyChange]leader pd2 frequently changed, leader-key is [/pd/7387036149936305935/leader]"]

2024-07-02T13:54:14.9189703Z [2024/07/02 13:53:31.833 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd1]

!!!resign pd-1 -> 0!!!
2024-07-02T13:54:14.9374977Z [2024/07/02 13:53:37.894 +00:00] [INFO] [member.go:358] ["try to resign etcd leader to next pd-server"] [from=pd1] [to=]
2024-07-02T13:54:14.9440845Z [2024/07/02 13:53:38.395 +00:00] [ERROR] [server.go:1718] ["campaign PD leader meets error due to etcd error"] [campaign-leader-name=pd1] [error="[PD:server:ErrLeaderFrequentlyChange]leader pd1 frequently changed, leader-key is [/pd/7387036149936305935/leader]"]


2024-07-02T13:54:14.9468215Z [2024/07/02 13:53:38.897 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd2]
2024-07-02T13:54:14.9755943Z [2024/07/02 13:53:44.934 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd2]
2024-07-02T13:54:15.0006334Z [2024/07/02 13:53:50.978 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd2]

!!!resign pd-2 -> 0!!!
2024-07-02T13:54:15.0132961Z [2024/07/02 13:53:57.018 +00:00] [INFO] [member.go:358] ["try to resign etcd leader to next pd-server"] [from=pd2] [to=]
2024-07-02T13:54:15.0152409Z [2024/07/02 13:53:57.520 +00:00] [ERROR] [server.go:1718] ["campaign PD leader meets error due to etcd error"] [campaign-leader-name=pd2] [error="[PD:server:ErrLeaderFrequentlyChange]leader pd2 frequently changed, leader-key is [/pd/7387036149936305935/leader]"]

2024-07-02T13:54:15.0162553Z [2024/07/02 13:53:58.021 +00:00] [INFO] [server.go:1738] ["campaign PD leader ok"] [campaign-leader-name=pd1]
```

--- a/Makefile
+++ b/Makefile
@@ -254,7 +254,7 @@ basic-test: install-tools
 
 ci-test-job: install-tools dashboard-ui pd-ut
 	@$(FAILPOINT_ENABLE)
-	./scripts/ci-subtask.sh $(JOB_COUNT) $(JOB_INDEX) || { $(FAILPOINT_DISABLE); exit 1; }
+	./scripts/ci-subtask.sh $(JOB_INDEX) || { $(FAILPOINT_DISABLE); exit 1; }
 	@$(FAILPOINT_DISABLE)
 
 TSO_INTEGRATION_TEST_PKGS := $(PD_PKG)/tests/server/tso

--- a/scripts/ci-subtask.sh
+++ b/scripts/ci-subtask.sh
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-# ./ci-subtask.sh <TOTAL_TASK_N> <TASK_INDEX>
+# ./ci-subtask.sh <TASK_INDEX>
 
 ROOT_PATH_COV=$(pwd)/covprofile
 ROOT_PATH_JUNITFILE=$(pwd)/junitfile

--- a/tests/server/region_syncer/region_syncer_test.go
+++ b/tests/server/region_syncer/region_syncer_test.go
@@ -234,11 +234,13 @@ func TestPrepareChecker(t *testing.T) {
 	re.NoError(err)
 	err = pd2.Run()
 	re.NoError(err)
+	re.NotEmpty(cluster.WaitLeader())
 	// waiting for synchronization to complete
 	time.Sleep(3 * time.Second)
+	leaderServer = cluster.GetLeaderServer()
 	err = cluster.ResignLeader()
 	re.NoError(err)
-	re.Equal("pd2", cluster.WaitLeader())
+	re.NotEqual(leaderServer.GetServer().Name(), cluster.WaitLeader())
 	leaderServer = cluster.GetLeaderServer()
 	rc = leaderServer.GetServer().GetRaftCluster()
 	for _, region := range regions {
@@ -282,16 +284,22 @@ func TestPrepareCheckerWithTransferLeader(t *testing.T) {
 	re.NoError(err)
 	err = pd2.Run()
 	re.NoError(err)
+	re.NotEmpty(cluster.WaitLeader())
 	// waiting for synchronization to complete
 	time.Sleep(3 * time.Second)
+	leaderServer = cluster.GetLeaderServer()
 	err = cluster.ResignLeader()
 	re.NoError(err)
-	re.Equal("pd2", cluster.WaitLeader())
+	re.NotEqual(leaderServer.GetServer().Name(), cluster.WaitLeader())
+	rc = cluster.GetLeaderServer().GetRaftCluster()
+	re.True(rc.IsPrepared())
 
-	// transfer leader to pd1, can start coordinator immediately.
+	// transfer leader, can start coordinator immediately.
+	leaderServer = cluster.GetLeaderServer()
 	err = cluster.ResignLeader()
 	re.NoError(err)
-	re.Equal("pd1", cluster.WaitLeader())
+	re.NotEqual(leaderServer.GetLeader().GetName(), cluster.WaitLeader())
+	rc = cluster.GetLeaderServer().GetServer().GetRaftCluster()
 	re.True(rc.IsPrepared())
 	re.NoError(failpoint.Disable("github.com/tikv/pd/pkg/schedule/changeCoordinatorTicker"))
 }

